File: src/llm/openai-embedding.ts
import OpenAI from 'openai';
import { withBackoff } from '../pipeline/concurrency.js';
import { CliError } from '../utils/errors.js';
import { batchEmbed, validateProbeDimensions } from './embed-batch.js';
import type { EmbeddingProvider } from './embedding-provider.js';

const BATCH = 96;
const DEFAULT_MODEL = 'text-embedding-3-small';
const DEFAULT_DIMENSIONS = 1536;

export interface OpenaiEmbeddingOptions {
  apiKey: string;
  model?: string;
  /**
   * Vector width. When unset, the provider emits the model's native output
   * and declares 1536 dims. Override for non-1536 models (e.g.
   * `text-embedding-3-large` = 3072) or to request reduced output via the
   * API's `dimensions` parameter.
   */
  dimensions?: number;
}

export class OpenaiEmbeddingProvider implements EmbeddingProvider {
  readonly name = 'openai';
  readonly dimensions: number;
  private client: OpenAI;
  private model: string;
  private explicitDimensions: number | undefined;

  constructor(opts: OpenaiEmbeddingOptions) {
    this.client = new OpenAI({ apiKey: opts.apiKey });
    this.model = opts.model ?? DEFAULT_MODEL;
    this.dimensions = opts.dimensions ?? DEFAULT_DIMENSIONS;
    this.explicitDimensions = opts.dimensions;
  }

  private buildParams(
    input: string | string[]
  ): OpenAI.Embeddings.EmbeddingCreateParams {
    const params: OpenAI.Embeddings.EmbeddingCreateParams = {
      model: this.model,
      input,
    };
    if (this.explicitDimensions !== undefined) {
      params.dimensions = this.explicitDimensions;
    }
    return params;
  }

  async probe(): Promise<void> {
    let response: OpenAI.Embeddings.CreateEmbeddingResponse;
    try {
      response = await this.client.embeddings.create(this.buildParams('probe'));
    } catch (err) {
      throw wrapOpenaiError('OpenAI embedding probe failed', err);
    }
    validateProbeDimensions({
      actual: response.data?.[0]?.embedding.length,
      declared: this.dimensions,
      providerLabel: 'OpenAI',
      model: this.model,
      remediationHint: `Set GDRIVESCOPE_OPENAI_EMBEDDING_DIMENSIONS=${response.data?.[0]?.embedding.length} or pick a different model.`,
    });
  }

  embed(texts: string[]): Promise<number[][]> {
    return batchEmbed(texts, BATCH, async (batch) => {
      try {
        const response = await withBackoff(() =>
          this.client.embeddings.create(this.buildParams([...batch]))
        );
        return response.data.map((item) => item.embedding);
      } catch (err) {
        throw wrapOpenaiError('OpenAI embedding call failed', err);
      }
    });
  }
}

function wrapOpenaiError(prefix: string, err: unknown): CliError {
  if (err instanceof OpenAI.APIError) {
    return new CliError(
      `${prefix} (status ${err.status}): ${err.message}`,
      'EMBED_CALL_FAILED'
    );
  }
  const msg = err instanceof Error ? err.message : String(err);
  return new CliError(`${prefix}: ${msg}`, 'EMBED_CALL_FAILED');
}

File: src/llm/openai_embedding.rs
//! OpenAI embeddings. Ported from `src/llm/openai-embedding.ts`.

use async_trait::async_trait;
use serde::Deserialize;
use serde_json::json;

use super::embed_batch::{batch_embed, validate_probe_dimensions, ProbeOptions};
use super::embedding::EmbeddingProvider;
use crate::error::{CliError, ErrorCode};

const OPENAI_EMBEDDINGS_API: &str = "https://api.openai.com/v1/embeddings";
const BATCH: usize = 96;
const DEFAULT_MODEL: &str = "text-embedding-3-small";
const DEFAULT_DIMENSIONS: usize = 1536;

#[derive(Debug, Deserialize)]
pub(super) struct EmbeddingsResponse {
    pub data: Vec<EmbeddingItem>,
}

#[derive(Debug, Deserialize)]
pub(super) struct EmbeddingItem {
    pub embedding: Vec<f32>,
}

pub struct OpenaiEmbeddingProvider {
    http: reqwest::Client,
    api_key: String,
    model: String,
    dimensions: usize,
    explicit_dimensions: Option<usize>,
}

pub struct OpenaiEmbeddingOptions {
    pub api_key: String,
    pub model: Option<String>,
    pub dimensions: Option<usize>,
}

impl OpenaiEmbeddingProvider {
    pub fn new(opts: OpenaiEmbeddingOptions) -> Self {
        Self {
            http: reqwest::Client::new(),
            api_key: opts.api_key,
            model: opts.model.unwrap_or_else(|| DEFAULT_MODEL.to_string()),
            dimensions: opts.dimensions.unwrap_or(DEFAULT_DIMENSIONS),
            explicit_dimensions: opts.dimensions,
        }
    }

    fn build_body(&self, input: &serde_json::Value) -> serde_json::Value {
        let mut body = json!({ "model": &self.model, "input": input });
        if let Some(d) = self.explicit_dimensions {
            body["dimensions"] = json!(d);
        }
        body
    }

    async fn call(&self, body: &serde_json::Value) -> Result<EmbeddingsResponse, CliError> {
        let resp = self
            .http
            .post(OPENAI_EMBEDDINGS_API)
            .bearer_auth(&self.api_key)
            .json(body)
            .send()
            .await
            .map_err(|e| {
                CliError::new(
                    format!("OpenAI embedding call failed: {e}"),
                    ErrorCode::EmbedCallFailed,
                )
            })?;
        if !resp.status().is_success() {
            let status = resp.status();
            let text = resp.text().await.unwrap_or_default();
            return Err(CliError::new(
                format!("OpenAI embedding call failed (status {status}): {text}"),
                ErrorCode::EmbedCallFailed,
            ));
        }
        resp.json().await.map_err(|e| {
            CliError::new(
                format!("OpenAI embedding parse failed: {e}"),
                ErrorCode::EmbedCallFailed,
            )
        })
    }
}

#[async_trait]
impl EmbeddingProvider for OpenaiEmbeddingProvider {
    fn name(&self) -> &str {
        "openai"
    }
    fn dimensions(&self) -> usize {
        self.dimensions
    }

    async fn embed(&self, texts: &[String]) -> Result<Vec<Vec<f32>>, CliError> {
        batch_embed(texts, BATCH, |batch| async move {
            let body = self.build_body(&json!(batch));
            let resp = self.call(&body).await?;
            Ok(resp.data.into_iter().map(|d| d.embedding).collect())
        })
        .await
    }

    async fn probe(&self) -> Result<(), CliError> {
        let body = self.build_body(&json!("probe"));
        let resp = self.call(&body).await?;
        validate_probe_dimensions(ProbeOptions {
            actual: resp.data.first().map(|d| d.embedding.len()),
            declared: self.dimensions,
            provider_label: "OpenAI",
            model: &self.model,
            remediation_hint:
                "Set GDRIVESCOPE_OPENAI_EMBEDDING_DIMENSIONS or pick a different model.",
        })
    }
}
